cloud/services/networkinterfaces/service.go

```diff
@@ -41,18 +41,18 @@ type Service struct {
 	LoadBalancersClient   loadbalancers.Client
 	PublicIPsClient       publicips.Client
 	InboundNATRulesClient inboundnatrules.Client
-	ResourceSkusClient    resourceskus.Client
+	ResourceSKUCache      *resourceskus.Cache
 }
 
 // NewService creates a new service.
-func NewService(scope NICScope) *Service {
+func NewService(scope NICScope, skuCache *resourceskus.Cache) *Service {
```

Comment:
why not make SKUCache part of machine scope just like it's added to cluster scope already and add `ResourceSKUCache` to `NICScope`?

Reply:
we can't that with VMs and VMSS yet because they haven't been refactored but eventually we'll be able to use that for all of them (also they won't have access to clusterScope anymore after the refactor)

Reply:
another option is to add it to clusterDescriber and use the cluster scope SKUCache but I don't think I like that as much.

Reply:
IMO we should construct clients once and inject them into services from the reconciler to reduce duplication, instead of constructing new clients for every service.

The natural extension of what you described to solve that problem ends up with us putting every service in one of the scopes, which feels wrong. I'd rather see us invert control, instantiate all clients during reconcile, and then pass only the required ones down to the services which need them. 

I don't mind changing it back since I'm not tackling the rest right now, but I do think we should move towards passing these down rather than instantiating them everywhere.

LMK what you think

Reply:
> https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/783/files#diff-7c0af1847b5ee759aadb398ac16c4d36R83

Completely agree.

I'm not sure how what I'm suggesting doesn't accomplish that though. I'm saying we shouldn't add a new dependency on clusterScope in the azure machine reconciler as we're moving towards getting rid of those https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/783/files#diff-fb5deef0b083d17ab538518df2570d4bR56.

What about adding a new interface to MachineScope, similar to ClusterDescriber, for the SKUCache and passing the clusterScope SKU cache to the MachineScopeParams when we create a newMachineScope? That way it's only created once and all services use the same SKU cache but we can explicitly define which service scopes need it and mock it.

Reply:
It's in the same vein as https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/798#discussion_r456109279 which I just ran into

I think all of the construction of the service/clients and their params can be pulled up a layer. Construct all of the clients, inject them, and then for each object to be reconciled, produce the spec struct for the desired object in the top level reconciler, and pass it to the service.

This way, each service is only a wrapper for Azure clients, and the functions themselves become a bit more "pure" and simpler to test directly. Like I said, definitely bigger change than this PR

Reply:
weird comment timing -- maybe i'm just overthinking it, I can try and see

Reply:
how's that? bit of weirdness with references to cache at different layers, but at least it removes the dep on scope everywhere

Reply:
I think it's fine for now, might play around with it a bit with the VM refactor.